Experimental: try new "Widget Group" block to act as container for Widgets

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Context https://github.com/WordPress/gutenberg/issues/32723.

This approach tries adding a new block called "Widget Box". It's basically just a container to add:

1. A Widget title.
2. InnerBlocks (your Widget).

This mimics the original widget markup format.

It also comes with some additional helpful touches:

1. A template for the inner blocks to pre-add the heading block for the user.
2. A transform to allow users to automatically convert a single widget to be wrapped in a Widget Box.

~Note that this could complement https://github.com/WordPress/gutenberg/pull/33878 where we provide a block pattern~. We can just[ use a `transform`](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-transforms/) which is what I've done now.

Closes https://github.com/WordPress/gutenberg/issues/32723

## To to

- [x] Add a new block that wraps a "title" and any of the supported widget area inner blocks
- [x] Decide wether we want to support nested widget boxes
- [ ] Render the block serverside and account for `before_title`, `after_title` for the rendered title markup
- [ ] Iterate on the UX of the block so that the differences between the editor and the front end are not a surprise


## How has this been tested?

* Go to Widgets screen.
* Add Widget Box block.
* FIll out the heading.
* Add a Widget block of your choice.
* Repeat.
* Test on front end.

To compare against the original/legacy widget markup pattern:

* Download and activate "Classic Widgets" Plugin.
* Go to previous/old widgets screen.
* Add a Widget and give it a title.
* Disable Plugin.
* Return to block-based Widgets screen.
* Add a Widget that is the same as the one you added in the Classic Widgets.
* Click on the "Transform" menu in the block toolbar.
* Click "Widget Box" to transform your block to be wrapped in a Widget Box.
* Fill in title.
* Save/Publish.
* View frontend of site.
* Check markup is near enough identical between legacy and block based widget (when using Widget Box).

## Screenshots <!-- if applicable -->

Tested using TwentyFifteen. The aim is visual parity when trying to create an "Archives" Widget using the "Widget Box" block 

<img width="646" alt="Screen Shot 2021-08-05 at 09 56 37" src="https://user-images.githubusercontent.com/444434/128322368-46d34b69-eec9-4621-9654-7c1be725aef3.png">


https://user-images.githubusercontent.com/444434/128347001-b7cd3e18-2175-42c7-a75e-4f25d798f7b2.mp4

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Register new Widget box block for Widgets screen only

draganescu commented:
Excellent start. This block, if we want it to solve #32723 be saved in the expected markup where the title is the immediate child of the widget's wrapper and has the standard "title" class applied. I think this can be done through the block's save function, no need for server rendering.

getdave commented:
> Excellent start. This block, if we want it to solve #32723 be saved in the expected markup where the title is the immediate child of the widget's wrapper and has the standard "title" class applied. I think this can be done through the block's save function, no need for server rendering.

@draganescu I tested this already on the front end and it's doing what you describe. I used Twenty Fourteen.

<img width="627" alt="Screen Shot 2021-08-05 at 09 09 58" src="https://user-images.githubusercontent.com/444434/128315530-972d201a-ee5b-4eef-b539-aef8e914ea18.png">

Update: ah thought I see there is still a nested `<div>`. How do I get rid of that?

Can anyone advise on exactly what are the standard classes that should be added and where?

getdave committed:
Better mirror legacy markup for Widgets

getdave commented:
@draganescu I've updated this PR to lighten the DOM and better match the default markup provided by legacy Widgets. I tested with twentyfifteen and once saved the Legacy Widget version (created by using the Classic Widgets Plugin) and the Block Based version (using the new "Widget Box" block) look near identical. The only thing I cannot control is the markup of the `<aside>` as I believe this is Theme dependent - I could be wrong however.

getdave committed:
Remove hardcoded class as not needed

talldan commented:
Right now, I'm thinking the block seems like a better option than a pattern. At least a block is something that can continue to be improved whereas a pattern, once inserted, is kind of set in stone.

Some thoughts:

- I think this should be in the widgets package so that it can be used in customize-widgets too.
- The heading block might be overkill, it could just be a RichText driven by a block attribute, which would make it possible to show the heading value in List View. It'd also make List View a bit cleaner.
- Should it use the button inserter for its inner blocks?
- A quick way to wrap existing blocks with this block (a bit like the group option/transform) might be good, but could be a follow-up.
- The same problem exists with group, but it's still hard for users to tell whether they're editing inside the widget box.

getdave committed:
Move Widget Box block to widgets package and update registration

getdave committed:
Register block in Widget Customizer and allow to appear

getdave committed:
Utilise block appender style

getdave committed:
Remove unnecessary /edit subfolder

getdave committed:
Add transform for any block to allow it to be wrapped in a Widget Box

getdave commented:
@talldan Some good points - thanks! I've converted them into a Todo list:

- [x] I think this should be in the widgets package so that it can be used in customize-widgets too.
- [ ] The heading block might be overkill, it could just be a RichText driven by a block attribute, which would make it possible to show the heading value in List View. It'd also make List View a bit cleaner.
- [x] Should it use the button inserter for its inner blocks?
- [x] A quick way to wrap existing blocks with this block (a bit like the group option/transform) might be good, but could be a follow-up.
- [ ] The same problem exists with group, but it's still hard for users to tell whether they're editing inside the widget box.

The two items that are _not_ completed are more complex.

> The heading block might be overkill, it could just be a RichText driven by a block attribute, which would make it possible to show the heading value in List View. It'd also make List View a bit cleaner.

I don't fully understand this. I assumed we would want a sematic heading. Indeed this seems to be what most Themes expect and what is generated by legacy Widgets by default. What value would this have and also would it continue to allow a 1-click-to-mirror experience for those wishing to retain the legacy block HTML markup pattern?

> The same problem exists with group, but it's still hard for users to tell whether they're editing inside the widget box.

Yes this is a wider problem. I believe we added "click through" to reusable blocks and template parts. Should/could we explore that here?

getdave commented:
@ddryo pinging you here as you raised the original Issue. Would you be open to testing this out and letting me know whether is goes someway towards solving the problem you noted?

getdave commented on packages/widgets/src/blocks/widget-box/block.json:
```diff
@@ -0,0 +1,19 @@
+		}
+	},
+	"supports": {
+		"html": false,
```
Should we allow this? I guess we ought to. There's no harm right?

draganescu replied:
Well I would not allow HTML of blocks to be changed in any block in the entire widgets editor.

getdave commented on packages/widgets/src/blocks/widget-box/edit.js:
```diff
@@ -0,0 +1,35 @@
+		},
+		{
+			template: TEMPLATE,
+			renderAppender: InnerBlocks.ButtonBlockAppender,
```
If necessary we could create our own appender (probably in a follow up) which could be more specific about the fact that you should insert a Widget.

draganescu replied:
You can insert any supported block not just a widget.

getdave commented on packages/widgets/src/blocks/widget-box/edit.js:
```diff
@@ -0,0 +1,35 @@
+			className: 'wp-widget-box__inner-blocks',
+		},
+		{
+			template: TEMPLATE,
```
Instead of a template we could introduce a "placeholder" for the block. When first inserted it would have a title field into which you would type the title of your Widget (e.g. `Latest Posts`). Then when you click submit it would auto create the Widget Box with a `core/heading` block already inserted and populated.

Just a thought...

draganescu replied:
Placeholders are for when we cannot represent the empty state with blocks in a useful and intuitive way. E.g. we don't want an "upload image" block, we want an image block placeholder that allows us to upload. But for this block the template will allow direct editing so we _can_ represent the empty state just fine. We can add some placeholder text to the headline maybe?

talldan commented:
> I don't fully understand this. I assumed we would want a sematic heading. Indeed this seems to be what most Themes expect and what is generated by legacy Widgets by default. What value would this have and also would it continue to allow a 1-click-to-mirror experience for those wishing to retain the legacy block HTML markup pattern?

It should be pretty much the same, the RichText would render a `h2`, so should be pretty comparable to having a block. It would mean that the heading can't be removed or changed to another block and the heading value could be used for list view. The markup would be slightly different as there would be an extra wrapper around the inner blocks.

draganescu commented on packages/customize-widgets/src/filters/widget-box-parent-support.js:
```diff
@@ -0,0 +1,35 @@
+ * @param {string} name     the name of the block.
+ * @return {Object} the new block settings.
+ */
+function addListBlockClassName( settings, name ) {
```
the name of this does not match what it does I believe

getdave replied:
Oh whoops! Forgot to rename.

getdave replied:
Fixed.

getdave commented:
https://user-images.githubusercontent.com/444434/128357229-d3864bc3-a2e5-45f8-ab2c-e84f03a83b9c.mp4

I gave RichText a go and this is what I found. Basically it somewhat spoils the UX as it appears you cannot make it a single line input like heading.

I assume the purpose is mainly so we can grab the "title" and potentially use it in the List View? 

I like that RichText can't easily be amended and it's cleans up the List View a little. The downside might be that now the user has less control over the output of the heading level. I'm not sure that's a benefit.

draganescu reviewed (commented):
This is cool! 

The one problem we have is that:

- the markup of the widget's title is `h2` by default
- it can be changed to be any tag and any class really via the `before_title` param to `register_sidebar` function
- therefore the title has to be stored in an attribute and then rendered serverside by taking the above into account

So to achieve the best compatible form for this maybe indeed a RichText connected to an attribute would work, particularly since we can't tell what markup it will have anyway and we can only approximate, so the `Heading` block's settings are only confusing (I set a H4 and I see the theme's H3 for instance).

getdave committed:
Initial attempt at RichText

getdave committed:
Rename misnamed method on filter

kevin940726 commented on packages/customize-widgets/src/filters/widget-box-parent-support.js:
```diff
@@ -0,0 +1,35 @@
+
+	// Remove the "parent" property.
+	// eslint-disable-next-line no-unused-vars
+	const { parent, ...newSettings } = settings;
```
An alternative would be

```suggestion
	const newSettings = { ...settings };
	delete newSettings.parent;
```

So that we don't need to disable eslint.

kevin940726 replied:
However, don't we also want to avoid nesting Widget Boxes here in the customizer too? I don't know if there's an official way to do that, but since that the _top-level block_'s name is `null`, seems like `parent: [ null ]` works. Not very readable though.

getdave replied:
> don't we also want to avoid nesting Widget Boxes here in the customizer too

Yes I'm just not sure how.

> but since that the top-level block's name is null, seems like parent: [ null ] works. Not very readable though.

Oh that `null` thing is strange. Is that reliable or a bit of a hack? 

I guess `InnerBlocks` has an `allowedBlocks` setting which we could use to include all blocks available int he Widgets Editor except `core/widget-box`. 

Perhaps we need a `disallowedBlocks` inverse setting on `InnerBlocks`?

tellthemachines replied:
I'm wondering if we should be bothered stopping people from nesting Widget Boxes at all. The worst that happens if they do is they get two headings one after the other, which is unsemantic, but is already possible to do with regular blocks if someone really wants to. Trying to prevent this might be more trouble than it's worth 😅

getdave replied:
If you feel this is ok then that's fine. I've removed the restriction thus simplifying the code.

kevin940726 commented on packages/customize-widgets/src/filters/widget-box-parent-support.js:
```diff
@@ -0,0 +1,35 @@
+	return newSettings;
+}
+
+addFilter(
```
It's just my preference, but I don't like using hooks in JavaScript 😅, it makes it very hard to trace code. Instead, could we just override the settings when we are registering the block type? It matches how the other registrations work in both editors too.

getdave replied:
> could we just override the settings when we are registering the block type?

Ah I see. Pass some arguments into the `registerWidgetBoxBlock()` method to choose how to configure it? Seems sensible. Happy to avoid filters!

getdave replied:
This restriction is now removed entirely on @tellthemachines's recommendation.

kevin940726 commented on packages/widgets/src/blocks/widget-box/edit.js:
```diff
@@ -0,0 +1,58 @@
+				tagName="h2"
+				aria-label={ __( 'Widget title' ) }
+				placeholder={ __( 'Add a Widget title' ) }
+				value={ attributes.widgetTitle }
```
Is it `widgetTitle` or is it `name`?

getdave replied:
I think `title` because that's the class used in the current widgets implementation. Also I think most legacy widgets use "title". I could be wrong. Why do you think `name` is more appropriate?

tellthemachines replied:
It should be `name` because that's what is declared in the attributes in `block.json`. Currently, when adding a title, saving and reloading the widgets editor, the block becomes invalid because the title isn't being saved.

getdave replied:
Yeh but that attribute is entirely arbitary. I just carried it across due to copy/paste 😄 

I believe it should be `title` (let's drop the `widget` prefix) because that is the terminology used in legacy Widget's (e.g. https://developer.wordpress.org/reference/hooks/widget_title/).

Is there a reason other than what I accidentally named the attr in`block.json` why `name` would be preferable?

getdave replied:
Updated to use `title` as the attribute.

kevin940726 commented on packages/widgets/src/blocks/widget-box/edit.js:
```diff
@@ -0,0 +1,58 @@
+
+					if ( isOriginal || value ) {
+						block = createBlock( 'core/heading', {
+							...attributes,
```
What other attributes do we want to spread here? If none then I think we can memoize this function (if necessary).

getdave replied:
Yeh this whole RichText area is subject to change. I kind-a copy/pasted from `core/heading` at this point. Needs refinement.

tellthemachines reviewed (commented):
Nice work! This seems more solid than a block pattern. 

Regarding the widget title markup, @draganescu raises a good point about themes having the ability to customise it. It's not only the element but also the classnames that can be customised, and that's why this changeset isn't working as expected on Twenty Twenty: in that theme, widget title styles are attached to a custom `heading-size-3` class. 

I'm not sure how (or even if) we can have access to what the theme defines in the `before_title` parameter of `register_sidebar`, but if we can get that, we'll need to use not only the heading level but also the classes.
 
Update: one other thing I noticed with using a RichText area for the heading is that, if the title is left empty, the heading markup is still output. That doesn’t seem too bad because if we really don’t want a title, there's no need to use a widget box, but thought I'd mention it anyway.

draganescu commented:
> I'm not sure how (or even if) we can have access to what the theme defines in the before_title parameter of register_sidebar, but if we can get that, we'll need to use not only the heading level but also the classes.

I don't thing we can get these in the editor, so the look of the widget box in the editor has to be obvious that it does not represent a preview but a structural draft, since the title is heavily changed by the theme on render (cc @critterverse)

tellthemachines commented:
> I don't thing we can get these in the editor, so the look of the widget box in the editor has to be obvious that it does not represent a preview but a structural draft, since the title is heavily changed by the theme on render

I might be missing something, but isn't the whole point of the widget box to allow adding titles similar to those rendered by the theme, because themes don't render titles for block widgets? If we can't know what heading level and/or classnames the theme is using for its legacy widgets, then we'll have to use the default `<h2 class="widget-title">` and accept that it won't work for every single theme out there (as it doesn't work for Twenty Twenty right now).

noisysocks commented:
We could maybe get to `before_title` and `after_title` if Widget Box is a dynamic block and we pass the sidebar ID from `WP_Block_Widget::widget()` to the block's `render_callback` via block context. It's a bit hacky but not _totally_ absurd.

talldan committed:
Add onReplace and identifier props to RichText in widget box block

talldan commented:
> I gave RichText a go and this is what I found. Basically it somewhat spoils the UX as it appears you cannot make it a single line input like heading.

I think it is possible. I think the missing bit is the `onReplace` prop, which is needed for `onSplit` to work. I pushed a commit with the changes. It's also possible to use the down arrow key to get to the appender.

Also added an `identifier` prop in that commit. I'm not sure if that's needed, but I remember seeing is used in block selection code. Can't hurt to add it!

getdave committed:
Improving self documenting code for allowing only single line for Widget title

getdave committed:
Remove restriction to limit Widget Box to being child of Widget Area

This is based on recommendation given in PR review

https://github.com/WordPress/gutenberg/pull/33881#discussion_r683864162

getdave committed:
Rename `widgetTitle` to `title` and update the block.json

getdave committed:
Update transform to convert preexisting heading blocks to the widget title prop

getdave commented:
The `RichText` based implementation in this block now has a bug whereby the Widget Box wrapper is removed when you hit "ENTER" when entering text into the widget title (the RichText).

1. Add Widget Box block.
2. Start adding a title.
3. Hit `ENTER`.
4. Open ListView - see Widget Box has been removed in favour of Heading + Paragraph.

This is due to the `onSplit` functionality which splits RichText into two separate blocks. However it doesn't not expect to be creating a new innerBlock which is what we'd want it to do.

This is a fundamental flaw and I'm not sure it's solvable without reengineering RichText itself.

getdave committed:
Allow richtext to split without overwritng the Widget Box

See https://github.com/WordPress/gutenberg/pull/33881#issuecomment-895126875

getdave commented:
> This is a fundamental flaw and I'm not sure it's solvable without reengineering RichText itself.

I've managed to fix this by providing an `onSplit` handler to `RichText` which programmatically updates the innerBlocks of the existing Widget Block (the one that contains the RichText) to contain a new paragraph block (plus any other existing innerBlocks). 

@talldan This seems to fix things.


https://user-images.githubusercontent.com/444434/128701776-42cea886-d26f-49a3-bd78-e09494db8a66.mp4

getdave commented:
> We could maybe get to `before_title` and `after_title` if Widget Box is a dynamic block and we pass the sidebar ID from `WP_Block_Widget::widget()` to the block's `render_callback` via block context. It's a bit hacky but not _totally_ absurd.

I'm currently stuck on making this block dynamic. I created a `index.php` and wired everything up but it doesn't seem to take effect. @noisysocks Any change you could take a look?

Also do we try to solve the `before_title` problem in this PR or in a followup? If in this PR then any advice on how to achieve that would be very much appreciated. I starting looking and got a bit lost.

critterverse commented:
This experiment introduces a new editing paradigm that falls somewhere between block editing and legacy widget editing in a way that I find confusing — for example, I would expect to be able to directly move/manipulate the title like a normal Heading block or else have the input look totally different like a legacy widget. My take is that despite the limitations with title styling, the [pattern approach](https://github.com/WordPress/gutenberg/pull/33878) feels like the more consistent and generally understandable solution 😅

tellthemachines commented:
> I'm currently stuck on making this block dynamic. I created a index.php and wired everything up but it doesn't seem to take effect. @noisysocks Any change you could take a look?

@getdave could you branch off from this branch, commit that experiment and push it to remote so we can have a play with it? That might also help us decide whether to do it here or as a separate PR. 

> This experiment introduces a new editing paradigm that falls somewhere between block editing and legacy widget editing in a way that I find confusing — for example, I would expect to be able to directly move/manipulate the title like a normal Heading block or else have the input look totally different like a legacy widget.

@critterverse By new paradigm, do you mean having both an editable area and an inner block container inside the block? We don't have any examples in Core yet, but the Navigation Link behaves like that when it has a submenu, and with #25940 the Gallery block will adopt that pattern too, so Widget Box definitely won't be the odd one out in the future. 

I think that if we can customise Widget Box with `before_title` and `after_title`, this is a clear winner over using a pattern, because then the titles will match with any theme, no matter what kind of markup it uses. If we can't, then it's less clear which is the best solution.

talldan commented:
I think the new paradigm is that it's half-block half-widget.

Just an idea, but I wonder if it should move more in the legacy widget direction visually to make clear that this is making a block act like a widget:
<img width="679" alt="Screenshot 2021-08-10 at 1 29 31 pm" src="https://user-images.githubusercontent.com/677833/128813456-182a1f47-4697-44d1-8f18-36d9a6440195.png">

getdave commented:
My only thought is I wonder if we're over complicating things with:

1. Rich Text
2. Attempting to utilise the `before_title` and `after_title` within the block.

Sometimes the best solutions are simple and flexible.

I still wonder whether using the Heading block instead of RichText (which is what this PR did initially) would afford more flexibility whilst still catering for 80% of use cases out of the box. If the heading level doesn't work for Theme "X" or "Y" then the user can simply toggle the heading level to match. Sure that's not _optimal_ - but it's a direct manipulation which is more in keeping with the rest of the editor.

Also related is the effort (not yet started) to cater for the `before_title` setting. This is going to make the block's behaviour quite "magical" as the markup won't be what is in the editor - rather it will be completely different on the front of the Theme. Whilst I appreciate we've established this is "ok" as a rule, in this case I feel it's better to have direct manipulation remain as king. 

Overall, I'm in favour of keeping things simple and flexible. Are there clear cut reasons why we _absolutely_ have to have the two features above to make the block viable?

> Just an idea, but I wonder if it should move more in the legacy widget direction visually to make clear that this is making a block act like a widget:

This is a route to explore. It certainly makes it more obvious that this "Widget Box" is a feature intended for backwards compatibility. 

I think I'm going to start branching off soon to explore various directions.

draganescu commented:
I think @talldan 's idea is the way forward. We really are over-complicating things if we want the sidebar settings to take effect in the block editor. There is no point in that. The title of the widget box can be a mere input, so that it does not look as interactive as to suggest what cofused @critterverse . We should make the block dynamic and respect the sidebar settings (before/after title) when we render this block.

@critterverse the pattern approach does not solve the problem as we don't have a way to render that pattern in a compatible manner where the markup of the heading and it's classes respect the theme's settings.

getdave commented:
> I think @talldan 's idea is the way forward.

I assume we mean [use the placeholder-style approach](https://github.com/WordPress/gutenberg/pull/33881#issuecomment-895742296)?

I'm in the middle of exploring that one now via a branch. Will push up shortly.

draganescu commented:
> I assume we mean use the placeholder-style approach?

yes! Placeholder! :)

getdave committed:
Attempt to make block dynamic

Note: this doesn't work. Why? I don't know.

getdave committed:
Register the WIdget Box server side

tellthemachines commented:
> Overall, I'm in favour of keeping things simple and flexible. Are there clear cut reasons why we absolutely have to have the two features above to make the block viable?

Using RichText mostly makes sense if we don't want the heading level and classnames to be editable by the user, which in turn makes sense if we're outputting them based on `before_title` and `after_title`, because we want the widget title to _exactly_ match the theme customisations for classic widgets. If we're going to use a Heading block here, then there's not much point to having a Widget Box instead of a pattern (unless it's easier for users to discover or something like that - which would be great to know, btw).

As for `before_title` and `after_title`, I think it's worth making a reasonable effort to get it working, because otherwise the Widget box won't work for all themes, and I suspect folks will keep opening issues about it. 

> We really are over-complicating things if we want the sidebar settings to take effect in the block editor. There is no point in that. 

@draganescu the issue here isn't so much having the block editor match the front end, it's having the front end output of the Widget box title match the titles generated by the theme for classic widgets. The front end is the thing that's not working yet 😅 

The placeholder discussion is a parallel issue. I'm fine with either solution tbh, but I'm still not clear on 

> I think the new paradigm is that it's half-block half-widget.

@talldan why is it half-widget? Is it because it mixes RichText with inner blocks? Or is it something else?

tellthemachines committed:
Render inner blocks on server side

tellthemachines commented:
> We could maybe get to `before_title` and `after_title` if Widget Box is a dynamic block and we pass the sidebar ID from `WP_Block_Widget::widget()` to the block's `render_callback` via block context.

I'm not at all clear on how this should work, could you elaborate a bit more @noisysocks ? Or anyone else?  

To pass the sidebar ID via block context, doesn't that require a parent block? I'm assuming widget area would be the parent in the editor, but what about the customizer?

talldan commented:
> @talldan why is it half-widget? Is it because it mixes RichText with inner blocks? Or is it something else?

Partly. When I look at it I don't think it's clear what the block does. In the front-end this is going to widgetize my block, but when editing it just looks like any other block and feels like this is how my content will appear in the front-end.

Making it similar to the legacy widget might make it more apparent. At least if the title is similar it makes it clearer that I'm editing a widget title. Adding some info text in the placeholder could also help.

noisysocks commented:
Apologies, incoming brain dump.

Note that in a new installation, WordPress creates regular Group and Heading blocks. There's no `.widget-title` class in the Heading blocks. See here:

https://github.com/WordPress/wordpress-develop/blob/9b565f7907ca1d4cb32eccc980de98683758df17/src/wp-admin/includes/upgrade.php#L405-L415

I think this is good. We ought to push for themes to style blocks in the widget area the same as blocks elsewhere in a site. Where a block is rendered should have no bearing on how it looks—that goes against the whole direct manipulation / "WYSIWYG" thing we've got going on. For example, Twenty Twenty-one was updated to not use `.widget-title` here:

https://github.com/WordPress/wordpress-develop/commit/8add05ff150f148d3a5b9f88ce12c05d5cc48897#diff-f03faeceffecb9471d071ec9885f73b7751809c5094922a86e34f38d3d659a11

Creating a group of blocks with a heading at the top should have exactly the same UX in the widgets editor as it does in the site editor—what @critterverse says in https://github.com/WordPress/gutenberg/pull/33881#issuecomment-895617409. This teaches users valuable skills which they can also use in the site editor.

I also think that we also should be thinking about how to make widget areas have no extraneous `.widget` markup. They should optionally behave like template areas. This should ease migration to full site editing. I fear Widget Box moves us in the wrong direction here by mixing the legacy widget paradigm with the new block paradigm.

So, what are the problems we need to solve? I believe there are two:

1. In a new WordPress installation, you get a bunch of groups with headings configured for you in your widget areas. This works great. But it is really unclear to users how you are supposed to add a _new_ group with a heading. 

2. It's not possible to insert a widget using the new block editor that renders _exactly_ the way that it always did. For example, if I want to insert an image with a title above it, and my theme hasn't been updated to play nicely with blocks, there is no escape hatch. I'm forced to install the Classic Widgets plugin.

I don't think we need to figure out how to add titles to blocks in a backwards compatible way. Again, this is mixing two paradigms. Legacy Widget can be used to insert a widget with a title. Blocks don't have titles. https://github.com/WordPress/gutenberg/issues/32723 can be fixed by allowing themes to opt out entirely from having blocks wrapped with the `<div class="widget">` wrapper. See https://github.com/WordPress/gutenberg/issues/32723#issuecomment-865144292.

So! My preference is to pursue three changes:

1. Go ahead with https://github.com/WordPress/gutenberg/pull/33878. I think I saw a mockup by @critterverse which had the block pattern appear in the quick inserter? I really liked that.

2. Have [`widget_types_to_hide_from_legacy_widget_block`](https://github.com/WordPress/wordpress-develop/blob/9b565f7907ca1d4cb32eccc980de98683758df17/src/wp-includes/block-editor.php#L249) be an empty array by default. We are being too prescriptive here. We ought to allow users to insert any legacy widget if they need to without having to do the nuclear option of installing Classic Widgets.

3. Pursue the idea raised in https://github.com/WordPress/gutenberg/issues/32723#issuecomment-865144292 and allow themes to opt out of the `<div class="widget">` wrapper around blocks. Perhaps this can be done when the sidebar is registered.

   ```php
   	register_sidebar(
   		array(
   			'name'          => esc_html__( 'Footer', 'twentytwentyone' ),
   			'id'            => 'sidebar-1',
   			'description'   => esc_html__( 'Add widgets here to appear in your footer.', 'twentytwentyone' ),
   			// Opt out of the .widget markup around blocks. The default is `true` which makes each block behave like a widget with no title.
   			'wrap_blocks'   => false,
   			// Only used in Legacy Widget.
   			'before_widget' => '<section id="%1$s" class="widget %2$s">',
   			'after_widget'  => '</section>',
   			'before_title'  => '<h2 class="widget-title">',
   			'after_title'   => '</h2>',
		)
	);

getdave commented:
I have no strong feeling either way. 

I do really like the idea of patterns (indeed that's [the approach I pursued first](https://github.com/WordPress/gutenberg/pull/33878)) but the fundamental flaw is that the markup is significantly different (additional inner HTML elements on group block) and this breaks the CSS of many Themes - including Core Themes).

I go into more detail on this problem here https://github.com/WordPress/gutenberg/pull/33878#issuecomment-893292806.

I notice that the Group block will omit the inner div _if_ the Theme supports the new Layout mode.

https://github.com/WordPress/gutenberg/blob/fce770ab877427ed5d1d8f1b0a5c46fca3c59172/packages/block-library/src/group/edit.js#L66

Perhaps we could also add a flag to be able to remove that inner div via a block attribute?

Unless we can solve that problem then patterns will remain a nice _conceptual_ idea that completely falls over when it collides with the reality of Themes.

If we cannot solve the problem with patterns then I still think "Widget Box" is our best option.

draganescu commented:
In https://github.com/WordPress/gutenberg/pull/33881#issuecomment-896495072 @noisysocks outlines a great summary of the problem.

I would like to stress out the point that the problem to solve is a **user problem**, not a developer problem. 

> my theme is setup in such a way that, in order for my sidebars to have some structure and hierarchy, I need to use titles between the sidebar's content.

- by generally enabling blocks to be added to the sidebar, we crippled the possibility of creating such a structure.
- heading blocks, even inside groups, don't account for the lack of titles because themes expect to own the styling of these elements.
- having access to all legacy widgets is fine, but if I want to add a block which does not have a widget equivalent I cannot make it conform to the structure and hierarchy of my sidebars and / or footer.

Pushing themes to update the way they style sidebars will _not_ solve users' problems for a long time. We may have been able to update a core theme, but I don't think other themes will rush to fix something they did not break. Allowing themes to opt out of widget wrapper for blocks may or may not lead to any observable result for users.

**Blocks don't have titles.** 

That is the main reason why, when we add blocks in a widget context, aka the widget editor, we need to provide a way to add titles to blocks. I still believe the widget box block is a very good path to get the best of both worlds:

- legacy widgets support titles
- blocks don't support titles
- if you want a block but you need a title, use the widget box block

When these blocks are migrated we can split them in heading + block.

noisysocks committed:
Rename to Widget Group, render before_title and after_title dynamically

noisysocks commented:
### I made some changes

I pushed up 6db4b04 which:

- Renders the theme's `before_title` and `after_title` dynamically by hooking into `before_dynamic_sidebar`.
- Removes the button block appender. This means there's not a tonne of ➕ icons in the editor once you have a few of these blocks set up.
- Renames the title prompt to _Add title_. This matches the post editor.
- Changes the title of the block to _Widget Group_... since this is a Group block.... but for widgets..... 🤷‍♂️

### Thoughts on the pattern approach

I also experimented with the pattern approach [over here](https://github.com/WordPress/gutenberg/pull/33878#issuecomment-897290992) and discovered that it only lets us solve this problem:

> In a new WordPress installation, you get a bunch of groups with headings configured for you in your widget areas. This works great. But it is really unclear to users how you are supposed to add a new group with a heading.

And not this one:

> my theme is setup in such a way that, in order for my sidebars to have some structure and hierarchy, I need to use titles between the sidebar's content.
> heading blocks, even inside groups, don't account for the lack of titles because themes expect to own the styling of these elements.

So, I think the pattern approach is off the table.

### Leaky abstraction

One thing I dislike about this approach is that our block editor abstraction is leaky. We're trying to add widget titles to the _block layer_ when widget titles have always existed in the _widget layer_.

To illustrate, here's what a search widget in a footer looked like in WordPress versions prior to 5.8:

```
+---------------------------------------+
| Sidebar                               |
|  id: footer                           |
| +-----------------------------------+ |
| | Widget                            | |
| |   id_base: search                 | |
| |   title: My search                | |
| +-----------------------------------+ |
+---------------------------------------+
```

Here's what a search block in a footer can look like in 5.8:

```
+---------------------------------------+
| Sidebar                               |
|  id: footer                           |
| +-----------------------------------+ |
| | Widget                            | |
| |   id_base: block                  | |
| | +-------------------------------+ | |
| | | Block                         | | |
| | |  name: core/search            | | |
| | +-------------------------------+ | |
| +-----------------------------------+ |
+---------------------------------------+
```

So how do we add a title to the search block? Here's what we are proposing here with Widget Box:

```
+---------------------------------------+
| Sidebar                               |
|  id: footer                           |
| +-----------------------------------+ |
| | Widget                            | |
| |   id_base: block                  | |
| | +-------------------------------+ | |
| | | Block                         | | |
| | |  name: core/widget-box        | | |
| | |  title: My search             | | |
| | | +---------------------------+ | | |
| | | | Block                     | | | |
| | | |  name: core/search        | | | |
| | | +---------------------------+ | | |
| | +-------------------------------+ | |
| +-----------------------------------+ |
+---------------------------------------+
```

But it would be more natural, I think, for the title to exist in the widget as it always has:

```
+---------------------------------------+
| Sidebar                               |
|  id: footer                           |
| +-----------------------------------+ |
| | Widget                            | |
| |   id_base: block                  | |
| |   title: My search                | |
| | +-------------------------------+ | |
| | | Block                         | | |
| | |  name: core/search            | | |
| | +-------------------------------+ | |
| +-----------------------------------+ |
+---------------------------------------+
```

Perhaps `core/widget-box` should work similarly to `core/legacy-widget` in that, during save, we convert it to a widget instead of serialising it as a block.

### Mixed paradigms

Another thing that bugs me is that some widget-specific markup (`.widget-title`) appears in the block editor when other widget-specific markup (`.widget`) does not.

Similarly, some things (the blocks inside a Widget Box) appear exactly as it does on the frontend because of editor styles but other things (the title of the Widget box) do not. (This is I think what @talldan was getting at in https://github.com/WordPress/gutenberg/pull/33881#issuecomment-895742296.)

I think that the dichotomy here is that there are two ways to think about the widgets editor:

1. **The widgets editor shows me my sidebars. Within each sidebar, I can add widgets. My theme decides how to present each sidebar and each widget.**
   
   This is the classic interpretation that the old widgets screen presented pretty clearly. APIs such as `before_widget`, `after_widget`, `before_title` and `after_title` imply that this interpretation is correct.

2. **The widgets editor shows me my sidebars. Within each sidebar, I can add content. I have complete control of how this content looks on my site's frontend.**
   
   This is the newer interpretation that aligns more closely with full site editing. Presenting a freeform block editor with editor styles for each sidebar implies that this interpretation is correct.

Perhaps we should allow the theme to pass `'mode' => 'classic'` or `'mode' => 'freeform'` into `register_sidebar()`. In classic mode, Widget Box is the primary mechanism for inserting content. In freeform mode, each sidebar behaves like a template part.

talldan commented:
> One thing I dislike about this approach is that our block editor abstraction is leaky. We're trying to add widget titles to the block layer when widget titles have always existed in the widget layer.

True, though a widget group could also just be like any other block if the `before_title` / `after_title` feature works in a way that degrades gracefully. And maybe a prominent transform to group/heading could be offered if there's a way to detect if the block is ever added in a non widget editor context.

But I think the main thing is that it's an optional feature, and a lot of themes won't really need this block, it's really just for themes that have a distinct style for widgets (gives them an outline or something) or has widget areas that behave in a particular way (like the columns layout in Twenty Twenty One).

Ideally the theme will be leaning towards a more freeform style.

noisysocks commented:
> True, though a widget group could also just be like any other block if the before_title / after_title feature works in a way that degrades gracefully.

I suppose 😅 At any rate, right now the block is in `@wordpress/widgets` so it's not a problem (yet).

> But I think the main thing is that it's an optional feature, and a lot of themes won't really need this block, it's really just for themes that have a distinct style for widgets (gives them an outline or something) or has widget areas that behave in a particular way (like the columns layout in Twenty Twenty One).
>
> Ideally the theme will be leaning towards a more freeform style.

Do you think the default widgets created in a new WordPress installation should continue to use regular groups?

noisysocks commented:
I think the next step is to explore the idea raised in https://github.com/WordPress/gutenberg/pull/33881#issuecomment-895742296. To re-iterate that conversation, if we make Widget Group look more akin to Legacy Widget then that should hopefully:

- Convey that this block is a useful tool for backwards compatibility and not something that users should reach first for. 
- Avoid confusion about why things look different in the editor (which does not have `.widget` markup, `before_title`, `after_title`, etc.) compared to the frontend.
- Avoid confusion about why there are two "Group" blocks that work differently.

critterverse commented:
Hey all, I'm also on board with @talldan's [suggestion](https://github.com/WordPress/gutenberg/pull/33881#issuecomment-895742296) to leverage the legacy widget UI. Just thinking out loud but I'm curious if there’s any way for the widget group block to be combined with the existing legacy widget block, or whether it needs to be a new separate block?

I would have to think through the UX some more but it might be something like a one-time set up placeholder that would either result in the usual legacy widget format (switches between "preview" and “editing" mode) or the something like the block/widget hybrid tried in this PR, depending on what you insert.

<img width="699" alt="legacy-widget-box-block" src="https://user-images.githubusercontent.com/77359364/129373472-6107518e-3d9d-4dbf-a488-7f9223afabee.png">

Let me know if this seems worth exploring — seems like it might be helpful to consolidate the "old way" into one legacy block and would help with some of the goals/considerations @noisysocks listed above.

noisysocks commented:
> Just thinking out loud but I'm curious if there’s any way for the widget group block to be combined with the existing legacy widget block, or whether it needs to be a new separate block?

Hmm. I'm not really sure. It might be worth running it by others on the design team. I lean towards a seperate block, because:

- More discoverable, e.g. users can search "group" in the inserter to find the block.
- More extendable, e.g. a theme could use block templates to make it so that only legacy widgets and widget groups are allowed at the top level. (We don't currently support block templates for widgets but we should.)

critterverse commented:
Hi all, sharing a first pass at the editing flow for the Widget Group block. I tried making the container block act like a Legacy Widget as much as possible to leverage our two existing editing flows:

https://user-images.githubusercontent.com/77359364/130693891-e8883e4a-6a1f-4882-ba56-7cbc91fa102c.mov

- **:00** Start with a one-time placeholder/set up block, similar to the set up flow for a Legacy Widget
- **:04** Within the placeholder, you can use the appender to open the quick inserter
- **:07** Once the block is inserted, a WYSIWYG-like preview of the title and block lockup is displayed. Focus is automatically shifted to the inner block upon insertion, with the Widget Group icon visible as a parent selector. Note that when the inner block is selected, you can see an outline around the parent block (probably a solid outline vs. dotted as shown above)
- **:11** When the Widget Group parent block is selected, a legacy widget “editing mode” is displayed, with an input form for editing the title
- **:15** When the block is deselected, it appears as a WYSIWYG-like preview

There are some weird things about this but I think that's going to be the case with any of the directions we take here. Thoughts?

noisysocks commented:
Thanks @critterverse! I think that mostly makes sense, and you're right about anything we do here being a bit awkward 😅

> :00 Start with a one-time placeholder/set up block, similar to the set up flow for a Legacy Widget

What do you think about ditching the "Add a block" label? We're very consistent about ➕ representing "add a block" so I think this label might be unnecessary.

> Once the block is inserted, a WYSIWYG preview of the title and block lockup is displayed.

This isn't possible unfortunately. We can display the title in a manner of our choosing but we can't make it appear _exactly_ as it will on the frontend because we cannot use the theme's `before_title` and `after_title` configuration in the editor.

critterverse commented:
>What do you think about ditching the "Add a block" label?

This makes sense! So more like this:

<img width="699" alt="widget-group-placeholder" src="https://user-images.githubusercontent.com/77359364/130799848-2ec06b60-c94e-44a5-9ee0-94dc2e8771ba.png">

>This isn't possible unfortunately. We can display the title in a manner of our choosing but we can't make it appear exactly as it will on the frontend because we cannot use the theme's before_title and after_title configuration in the editor.

Sorry, WYSIWYG was bad wording on my part — I just mean that in its resting state, the Widget Group can display like a Legacy Widget in "preview mode" (no worries about the title not displaying exactly as it will on the front end).

getdave commented:
It's great to have a clearer direction here ✨ 

My initial thoughts are that it might not be obvious that you have to click the parent `Widget Group` block to be able to edit the title again. 

One thing I was thinking was that the core team are working on [the ability for a child to consume the parents toolbar controls](https://github.com/WordPress/gutenberg/pull/33955) so perhaps we could even make the title part of the Widget Group's block toolbar and then allow the child blocks to display that title control if they are children of the Widget Group. However, I may be overcomplicating things...

Looking forward to seeing this evolve.

critterverse commented:
>we could even make the title part of the Widget Group's block toolbar and then allow the child blocks to display that title control if they are children of the Widget Group. However, I may be overcomplicating things...

This does seem a bit complicated, both from the perspective of treating the title like a toolbar action and tying the legacy UI to that (if I’m imagining this correctly). Interesting PR though, glad to have it on my radar!

>My initial thoughts are that it might not be obvious that you have to click the parent Widget Group block to be able to edit the title again.

Here's an alt version of the above flow that might help with this:

https://user-images.githubusercontent.com/77359364/131010050-8c978bc0-1fc2-47d0-b8dd-eb0620f4c5ac.mov

In this version the title input is removed from the placeholder set up and when a block is inserted, the editing interface for the Widget Group is displayed rather than first editing the inner block. This is slightly closer to the normal Legacy Widget set up flow and could help make the title more discoverable. 

Note that I changed the legacy editing UI to match the Legacy Widget block more closely, with the inner block type displayed at the top rather than "Widget Group." I think this makes the most sense? Similar to Legacy Widgets, we could use the block type as the default title to display if the title input field is left blank. This would work nicely for most widget blocks like "Archives" but would be kinda weird if you insert a Cover block or something 🤔

noisysocks commented:
> Note that I changed the legacy editing UI to match the Legacy Widget block more closely, with the inner block type displayed at the top rather than "Widget Group." I think this makes the most sense? Similar to Legacy Widgets, we could use the block type as the default title to display if the title input field is left blank. This would work nicely for most widget blocks like "Archives" but would be kinda weird if you insert a Cover block or something 🤔

How would this work with inserting multiple blocks into the Widget Group, though? Or would you have to insert a Group block first?

noisysocks commented:
I had an idea! Please tell me if it's terrible.

We already have a design metaphor for this problem of "how do you present _visual content_ (that looks as it does on the frontend) within an _abstract container_ (that does **not** necessarily look as it does on the frontend)": the widget area itself.

<img width="720" alt="Screen Shot 2021-08-27 at 13 32 34" src="https://user-images.githubusercontent.com/612155/131067741-414a3ab7-8acc-4434-9cee-b3cc2583266a.png">

Perhaps Widget Group could borrow this? Here's my attempt at butchering @critterverse's lovely Figma file.

![Widget Box](https://user-images.githubusercontent.com/612155/131070539-be967499-1027-4234-bd38-1dddb4bff7ad.png)

What I don't like:

- I don't know how to make the title look editable. Perhaps it should be a text field.
- It's a little "inception"-ey having the exact same UI element repeated within itself. Perhaps we can do more to visually distinguish the two UI elements while keeping the core metaphor present.

What I do like:

- There's a consistency and clear separation within the interface between UI elements that are grey which represent _abstract content_ and UI elements that are green (in Twenty Twenty-one) which represent _visual content_.

critterverse commented:
Interesting exploration @noisysocks, but I agree about this implementation being a bit inception-y and I'm not sure I completely follow the design metaphor — for example in this scenario I would expect "Footer" (or whatever widget area title) to be displayed on the front end in the same way as "Archives." 

I prefer the legacy UI approach because:

- it seems helpful/important to convey that this is for backwards compatibility
- it builds upon the two existing editing flows we already have

I can keep looking into what might happen if users add or change inner blocks if we think we're on the right track with [this direction](https://github.com/WordPress/gutenberg/pull/33881#issuecomment-906658061).

noisysocks commented:
> for example in this scenario I would expect "Footer" (or whatever widget area title) to be displayed on the front end in the same way as "Archives."
> it seems helpful/important to convey that this is for backwards compatibility

Yeah fair enough 🙂

> I can keep looking into what might happen if users add or change inner blocks if we think we're on the right track with this direction.

I _think_ I prefer the first prototype (https://github.com/WordPress/gutenberg/pull/33881#issuecomment-905025323) over the second (https://github.com/WordPress/gutenberg/pull/33881#issuecomment-906658061) because: 

- In the second, when you select the block ("Archives"), it's not clear where the block goes. I think that we've trained users to expect that clicking a block in the inserter takes them to that block.

- I don't think it's such a big deal to show "Widget Group" as the placeholder's heading. In a Legacy Widget block, the placeholder's heading does not necessarily have anything to do with the title input. For example here is an Image widget in a Legacy Widget block.
  
  https://user-images.githubusercontent.com/612155/131274420-13738c64-0665-4bfe-9279-aac01c31b357.mp4

- The blue dotted line that appears around the Widget Group is, I think, enough of a hint towards how one can edit the title.

---

As a "thought experiment", here's how a Text widget behaves when it's in a Legacy Widget block. This scenario is _slightly_ analogous to a Widget Group because Text widgets support adding multiple "blocks" (except we don't call them that) of content.

https://user-images.githubusercontent.com/612155/131274725-567d65ca-3b7c-4b2d-a300-5a931c237026.mp4

critterverse commented:
Thanks for these videos @noisysocks, both are helpful for thinking about this :)

>I think I prefer the first prototype... I don't think it's such a big deal to show "Widget Group" as the placeholder's heading

Using the above as a starting point, I'm still thinking that autofilling the title input based on the block that's initially inserted could work. The title can be manually updated if the block type isn't a particularly public-facing title or if you start changing/adding blocks (this doesn't seem too unexpected, imo).

I like the idea of leaving the title input empty and not showing the title by default but one concern is that it might make the title harder to discover. For example, this note would no longer apply if the default behavior is to only show the inner block:

>The blue dotted line that appears around the Widget Group is, I think, enough of a hint towards how one can edit the title.

noisysocks commented:
> I like the idea of leaving the title input empty and not showing the title by default but one concern is that it might make the title harder to discover. For example, this note would no longer apply if the default behavior is to only show the inner block:
> 
> > The blue dotted line that appears around the Widget Group is, I think, enough of a hint towards how one can edit the title.

Yeah really good point.

I'll start to implement one of the two flows (in a new PR) so that we can see what works in practice.

noisysocks commented:
I opened https://github.com/WordPress/gutenberg/pull/34484 which takes this PR and implements the flow in https://github.com/WordPress/gutenberg/pull/33881#issuecomment-906658061. Let's continue the discussion there. Thanks @getdave for working on this and getting the ball rolling 🙂
